src/Workspaces/Core/Portable/Utilities/BKTree.Builder.cs

```diff
@@ -0,0 +1,294 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Roslyn.Utilities
+{
+    internal partial class BKTree
+    {
+        private class Builder
+        {
+            // The number of edges we pre-allocate space for for each node in _compactEdges.
+            //
+            // To make the comments simpler below, i'll use '4' as a synonym for CompactEdgeAllocationSize.
+            // '4' simply reads better and makes it clearer what's going on.
+            private const int CompactEdgeAllocationSize = 4;
+
+            // Instead of producing a char[] for each string we're building a node for, we instead 
+            // have one long char[] with all the chracters of each string concatenated.  i.e.
+            // "foo" "bar" and "baz" becomes { f, o, o, b, a, r, b, a, z }.  Then in _wordSpans
+            // we have the text spans for each of those words in this array.  This gives us only
+            // two allocations instead of as many allocations as the number of strings we have.
+            //
+            // Once we are done building, we pass this to the BKTree and its nodes also state the
+            // span of this array that corresponds to the word they were created for.  This works
+            // well as other dependent facilities (like EditDistance) can work on sub-arrays without
+            // any problems.
+            private readonly char[] _concatenatedLowerCaseWords;
+            private readonly TextSpan[] _wordSpans;
+
+            // Note: while building a BKTree we have to store children with parents, keyed by the
+            // edit distance between the two.  Naive implementations might store a list or dictionary
+            // of children along with each node.  However, this would be very inefficient and would
+            // put an enormous amount of memory pressure on the system.
+            //
+            // Emperical data for a nice large assembly like mscorlib gives us the following 
+            // information:
+            // 
+            //      Unique-Words (ignoring case): 9662
+            // 
+            // For each unique word we need a node in the BKTree. If we stored a list or dictionary 
+            // with each node, that would be 10s of thousands of objects created that would then 
+            // just have to be GCed.  That's a lot of garbage pressure we'd like to avoid.
+            //
```

Comment:
Without looking deeper, do we have logic to ensure that the array we store wouldn't get enormously large? If arrays in the BKTree ended up on the LOH, I could see that being problematic for GC pressure as well.